### ISSUE TITLE
e2e flake: "Networking should function for intra-pod communication [Conformance]" failed in 2000-node GKE cluster

### ISSUE THREAD

**Issue by wojtek-t** [kubernetes/kubernetes#27039]
The failure is not really test-specific. It failed with:
```
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/networking.go:214 Expected error: <*url.Error | 0xc8222db470>: { Op: "Get", URL: "https://104.196.128.41/api/v1/watch/namespaces/e2e-tests-nettest-fwge9/pods?fieldSelector=metadata.name%3Dnettest-44a4q", Err: {}, } Get https://104.196.128.41/api/v1/watch/namespaces/e2e-tests-nettest-fwge9/pods?fieldSelector=metadata.name%3Dnettest-44a4q: net/http: TLS handshake timeout not to have occurred
```

Looking into events from the same test, I've see three other tls-handshake timeouts faced by kubelets:
```
Jun  8 03:01:04.774: INFO: At {2016-06-08 02:58:06 -0700 PDT} - event for nettest-ikqf7: {kubelet gke-gke-large-cluster-default-pool-1-03ee5a12-zzfu} FailedSync: Error syncing pod, skippin       g: Get https://104.196.128.41/api/v1/namespaces/e2e-tests-nettest-fwge9/secrets/default-token-6qc6e: net/http: TLS handshake timeout
Jun  8 03:01:04.774: INFO: At {2016-06-08 02:58:06 -0700 PDT} - event for nettest-xrqpl: {kubelet gke-gke-large-cluster-default-pool-1-03ee5a12-uqiu} FailedMount: Unable to mount volumes f       or pod "nettest-xrqpl_e2e-tests-nettest-fwge9(40c2d112-2d5f-11e6-9847-42010af00002)": Get https://104.196.128.41/api/v1/namespaces/e2e-tests-nettest-fwge9/secrets/default-token-6qc6e: net/       http: TLS handshake timeout
Jun  8 03:01:04.774: INFO: At {2016-06-08 02:58:06 -0700 PDT} - event for nettest-xrqpl: {kubelet gke-gke-large-cluster-default-pool-1-03ee5a12-uqiu} FailedSync: Error syncing pod, skippin       g: Get https://104.196.128.41/api/v1/namespaces/e2e-tests-nettest-fwge9/secrets/default-token-6qc6e: net/http: TLS handshake timeout
```

I think that we should simply retry "handshake timeout" errors.
@lavalamp @smarterclayton @krousey - WDYT?

@zmerlynn @gmarek @kubernetes/sig-api-machinery

**Comment by smarterclayton** [comment]
So maybe a counter point - should we not instead just set a longer timeout
if we are going to retry that?  After all, TLS handshake timeout could be
the server going away.  I would be ok with retrying Temporary() errors (as
returned by net.Error interface), but Timeout() seems a harder sell.

On Wed, Jun 8, 2016 at 8:02 AM, Wojciech Tyczynski <notifications@github.com

**Comment by wojtek-t** [comment]
That sounds fine to me too.

On the other hand, maybe some else bad is happening here. We already have TLShandshakeTimeout set to 10s. Do you think larger values make sense?

**Comment by ncdc** [comment]
I've long wondered why we're bumping up against the full 10s timeout in our tests. Surely something is seriously wrong somewhere (extremely overloaded?) if that is happening?

**Comment by wojtek-t** [comment]
Actually, the test failed the second time now (all other tests are passing in 2000-node cluster), currently with a different error:
```
07:25:54 • Failure [1853.935 seconds]
07:25:54 [k8s.io] Networking
07:25:54 /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:639
07:25:54   should function for intra-pod communication [Conformance] [It]
07:25:54   /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/networking.go:214
07:25:54 
07:25:54   Expected error:
07:25:54       <*url.Error | 0xc821bde150>: {
07:25:54           Op: "Post",
07:25:54           URL: "https://104.196.128.41/api/v1/namespaces/e2e-tests-nettest-3v2ez/pods",
07:25:54           Err: {
07:25:54               Op: "dial",
07:25:54               Net: "tcp",
07:25:54               Source: nil,
07:25:54               Addr: {
07:25:54                   IP: "\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xff\xffhĀ)",
07:25:54                   Port: 443,
07:25:54                   Zone: "",
07:25:54               },
07:25:54               Err: {},
07:25:54           },
07:25:54       }
07:25:54       Post https://104.196.128.41/api/v1/namespaces/e2e-tests-nettest-3v2ez/pods: dial tcp 104.196.128.41:443: i/o timeout
07:25:54   not to have occurred
07:25:54 
07:25:54   /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/networking.go:293
```

So most probably something else bad is happening here...

**Comment by wojtek-t** [comment]
Actually, now I recall that @thockin said something that services with many instances (500+) can cause problems. Here we have service with 2000 underlying pods, each pod on a different not.
@thockin - can this be related?

**Comment by zmerlynn** [comment]
Is this the first set of runs with all 2k nodes schedulable? In which case
that's probably what it is.

**Comment by zmerlynn** [comment]
@wojtek-t: The failure in https://github.com/kubernetes/kubernetes/issues/27039#issuecomment-224608419 looks like we flaked on trying to create the pod. Unfortunately, because of #26379, I don't have master logs, so I'm not sure what the master saw (if anything) there. Should we throw a retry in the e2e? It's kind of evil, but OTOH we're creating 2k pods.

**Comment by wojtek-t** [comment]
yes - before there were ~1000 routes when running this test

and yes - master log would be super useful. It ma be because of overloaded master, but... We've seen i/o timeout errors in small clusters too without any significant load - e.g. https://github.com/kubernetes/kubernetes/issues/26628

**Comment by wojtek-t** [comment]
But sure - overloaded apiserver is definitely a possible cause here.

**Comment by thockin** [comment]
The only real risk of lots of instances is overflowing the etcd size limits with the Endpoints struct, I think.  Since the change to iptables proxy, the iptables setup time is approximately O(1) rather than O(n) with the userspace.

**Comment by wojtek-t** [comment]
I will try to reproduce and debug it

**Comment by wojtek-t** [comment]
OK - I managed to reproduce it locally. So few findings are:
- the machine is relatively occupied, but we doesn't seem to be cpu-starved (we seem to be using up to 25 cores from 32 available)
- but my feeling is that we are network-limited

So the issue seem to be that we have 2000 pods, each of them, asking for endpoints of the service. The service contains 2000 pods, which means there are 2000 endpoints for the service.

It seems that master machine is simply network-starved.

I will try to switch "nettest" containers to use protobufs for communication with apiserver and see if that helps.

**Comment by wojtek-t** [comment]
The problem is generally that every of those 2000 pods is periodically sending "get endpoint" request to apiserver and this endpoint is almost 1MB of data.

Since those pods are sending get endpoint request every 5s, that means that every second ~400 pods are asking for 1MB of data each. This is a loooooooot of data that we need to send.

It basically seems that the test is not designed to run in large clusters - I will try to fix this test. Migrating webserver to protobufs didn't help that much.

**Comment by wojtek-t** [comment]
And to clarify a bit more: if we have say 1900 pods in a service, adding new pod results in changing the underlying "Endpoint" for this service. This in turns results in sending the new endpoint to all nodes (to kube-proxy). That basically means, that since Endpoint has ~1MB with ~2000 pods, adding a new pod results in sending ~2GB of data from apiserver to nodes.

So what I'm going to do is to limit the number of pods in the service in this test.

**Comment by wojtek-t** [comment]
Also, those "i/o timeout" failure confirm my theory, because sometime, we even not start handshaking (although apiserver is up all the time)

**Comment by dims** [comment]
@wojtek-t - This sounds like a scaling upper limit for 1.3 at least. is it?

**Comment by wojtek-t** [comment]
@dims - yes and no; basically even with 2000 pods in a service the cluster is working and your pods are working fine; the issue here is that:
- these pods are sending requests to apiserver (which is quite artificial setup)
- those requests are super expensive
But if you decide to retry the requests (those that are failing) you will succeed. There are just moments where network on the master machine is overloaded.

I can imagine the situation with overloaded network even without having any service. Let's say that you have a full 1000-node cluster with 30.000 pods running on them. And you call "List all pods" request. And you do it concurrently from 100 different computers. Probably - the same will happen, because that mean you need to send ~15GB of data to clients at the same time.

**Comment by wojtek-t** [comment]
In other words: if you have a pod that in a loop calls "LIST all pods", and you start 30.000 of them in 1000-node cluster (which is already supported configuration in 1.2) then those pods will also network-starve the apiserver. Does it mean we don't support 30.000 pods in 1.2? In my opinion it doesn't.

**Comment by dims** [comment]
@wojtek-t gotcha. Thanks for the detailed explanation.

**Comment by smarterclayton** [comment]
A 2000 node cluster is a weapon of mass internet destruction, and should be
treated with care.

**Comment by wojtek-t** [comment]
Completely agree with @smarterclayton 
BTW - we would like to optimize this at some point too.

**Comment by lavalamp** [comment]
@bgrant0607 has been saying for a while that endpoints needs to be sharded.

**Comment by lavalamp** [comment]
I think I wrote this test way back in the day. It was meant for 5 node clusters, not 2k node clusters. IIRC, it does basically n^2 connections, not even counting what the endpoints controller has to do. There's any number of ways to fix this. It just needs to test that Pod IP addresses work from every node.

**Comment by wojtek-t** [comment]
It does n^2 connections, but it most of all also stresses apiserver. Anyway - I've already sent out some quick fix out and we should fix that properly once I'm fully done with 1.3 work.

**Comment by bgrant0607** [comment]
@lavalamp @wojtek-t @smarterclayton Yes, this was a timebomb, which we've known from the beginning. Internally we shard by 100-endpoint groups to avoid this problem.

This is related to pagination (#2349), but that's not applicable, since that's about list chunking. This is a single key in etcd. Additionally, etcd's single-value size limit puts an absolute cap on service size currently.

We need a relatively stable sharding policy (e.g., by subset+CIDR) so that we don't repeatedly need to move endpoints between multiple keys. 

Of course, the only reason we need to persist endpoints at all is due to support for manually populated endpoints. If we replaced that by the component registration proposal (#18610), we could just store endpoints in memory.
